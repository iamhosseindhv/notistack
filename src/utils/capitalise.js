/**
 * Capitalises a piece of string
 * @param {string} text
 */
const capitalise = text => text.charAt(0).toUpperCase() + text.slice(1);

export default capitalise;
