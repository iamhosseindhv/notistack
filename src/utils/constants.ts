import { SnackbarClassKey } from '@material-ui/core/Snackbar';
import { CloseReason, ContainerClassKey, SnackbarProviderProps } from '../index';
import { SnackbarItemProps } from '../SnackbarItem';
import { Snack } from '../SnackbarProvider';

export const allClasses: {
    mui: Record<SnackbarClassKey, {}>;
    container: Record<ContainerClassKey, {}>;
} = {
    mui: {
        root: {},
        anchorOriginTopCenter: {},
        anchorOriginBottomCenter: {},
        anchorOriginTopRight: {},
        anchorOriginBottomRight: {},
        anchorOriginTopLeft: {},
        anchorOriginBottomLeft: {},
    },
    container: {
        containerAnchorOriginTopCenter: {},
        containerAnchorOriginBottomCenter: {},
        containerAnchorOriginTopRight: {},
        containerAnchorOriginBottomRight: {},
        containerAnchorOriginTopLeft: {},
        containerAnchorOriginBottomLeft: {},
    },
};

export const MESSAGES = {
    NO_PERSIST_ALL: 'WARNING - notistack: Reached maxSnack while all enqueued snackbars have \'persist\' flag. Notistack will dismiss the oldest snackbar anyway to allow other ones in the queue to be presented.',
};

export const SNACKBAR_INDENTS = {
    view: { default: 20, dense: 4 },
    snackbar: { default: 6, dense: 2 },
};

export const capitalise = (text: string): string => text.charAt(0).toUpperCase() + text.slice(1);

export const originKeyExtractor = (anchor: Snack['anchorOrigin']): string => (
    `${capitalise(anchor.vertical)}${capitalise(anchor.horizontal)}`
);

/**
 * Omit SnackbarContainer class keys that are not needed for SnackbarItem
 */
export const omitContainerKeys = (classes: SnackbarProviderProps['classes'] = {}): SnackbarItemProps['classes'] => (
    // @ts-ignore
    Object.keys(classes).filter(key => !allClasses.container[key]).reduce((obj, key) => ({ ...obj, [key]: classes[key] }), {})
);

export const REASONS: { [key: string]: CloseReason } = {
    CLICKAWAY: 'clickaway',
    MAXSNACK: 'maxsnack',
    INSTRUCTED: 'instructed',
};
