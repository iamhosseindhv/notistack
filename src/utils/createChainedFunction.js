/**
 * @link https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/utils/createChainedFunction.js
 */
export default function createChainedFunction(funcs, extraArg) {
    return funcs.reduce((acc, func) => {
        if (func == null) return acc;

        if (process.env.NODE_ENV !== 'production') {
            if (typeof func !== 'function') {
                // eslint-disable-next-line no-console
                console.error('Invalid Argument Type. must only provide functions, undefined, or null.');
            }
        }

        return function chainedFunction(...args) {
            const argums = [...args];
            if (extraArg && argums.indexOf(extraArg) === -1) {
                argums.push(extraArg);
            }
            acc.apply(this, argums);
            func.apply(this, argums);
        };
    }, () => { });
}
