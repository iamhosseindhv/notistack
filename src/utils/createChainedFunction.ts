import { SnackbarKey } from 'src/types';

const noOp = () => {
    /* */
};

/**
 * Credit to MUI team @ https://mui.com
 * Safe chained function.
 *
 * Will only create a new function if needed,
 * otherwise will pass back existing functions or null.
 */
export default function createChainedFunction<Args extends any[], This>(
    funcs: Array<((this: This, ...args: Args) => any) | undefined>,
    snackbarId?: SnackbarKey
): (this: This, ...args: Args) => void {
    // @ts-ignore
    return funcs.reduce((acc, func) => {
        if (func === null || func === undefined) {
            return acc;
        }

        return function chainedFunction(...args) {
            const argums = [...args] as any;
            if (snackbarId && argums.indexOf(snackbarId) === -1) {
                argums.push(snackbarId);
            }
            // @ts-ignore
            acc.apply(this, argums);
            func.apply(this, argums);
        };
    }, noOp);
}
