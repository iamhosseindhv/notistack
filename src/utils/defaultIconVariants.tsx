import React from 'react';
import SvgIcon, { SvgIconProps } from '@material-ui/core/SvgIcon';
import { IconVariant } from '../index';

type Icon = (props: SvgIconProps) => JSX.Element;

const CheckIcon: Icon = props => (
    <SvgIcon {...props}>
        <path d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2,
        4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,
        0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
        />
    </SvgIcon>
);

const WarningIcon: Icon = props => (
    <SvgIcon {...props}>
        <path d="M13,14H11V10H13M13,18H11V16H13M1,21H23L12,2L1,21Z" />
    </SvgIcon>
);

const ErrorIcon: Icon = props => (
    <SvgIcon {...props}>
        <path d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,
        20 12,20M12,2C6.47,2 2,6.47 2,12C2,17.53 6.47,22 12,22C17.53,22 22,17.53 22,12C22,6.47 17.53,
        2 12,2M14.59,8L12,10.59L9.41,8L8,9.41L10.59,12L8,14.59L9.41,16L12,13.41L14.59,16L16,
        14.59L13.41,12L16,9.41L14.59,8Z"
        />
    </SvgIcon>
);

const InfoIcon: Icon = props => (
    <SvgIcon {...props}>
        <path d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,
        12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,
        10 0 0,0 12,2M11,17H13V11H11V17Z"
        />
    </SvgIcon>
);

const iconStyles = {
    opacity: 0.9,
    fontSize: 20,
    marginInlineEnd: 8,
};

const defaultIconVariants: Omit<IconVariant, 'default'> = {
    success: <CheckIcon style={iconStyles} />,
    warning: <WarningIcon style={iconStyles} />,
    error: <ErrorIcon style={iconStyles} />,
    info: <InfoIcon style={iconStyles} />,
};

export default defaultIconVariants;
