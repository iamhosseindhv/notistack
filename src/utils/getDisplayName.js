// https://github.com/JamesMGreene/Function.name/blob/58b314d4a983110c3682f1228f845d39ccca1817/Function.name.js#L3
const fnNameMatchRegex = /^\s*function(?:\s|\s*\/\*.*\*\/\s*)+([^(\s/]*)\s*/;
const getFunctionName = (fn) => {
    const match = `${fn}`.match(fnNameMatchRegex);
    const name = match && match[1];
    return name || '';
};

/**
 * @param {function} Component
 * @param {string} fallback
 * @returns {string | undefined}
 */
const getFunctionComponentName = (Component, fallback = '') => (
    Component.displayName || Component.name || getFunctionName(Component) || fallback
);

const getWrappedName = (outerType, innerType, wrapperName) => {
    const functionName = getFunctionComponentName(innerType);
    return (
        outerType.displayName || (functionName !== '' ? `${wrapperName}(${functionName})` : wrapperName)
    );
};

/** 
 * From react-is
 * @link https://github.com/facebook/react/blob/master/packages/shared/ReactSymbols.js
 */
const ForwardRef = () => {
    const symbolFor = typeof Symbol === 'function' && Symbol.for;
    return symbolFor ? symbolFor('react.forward_ref') : 0xead0
}

/**
 * https://github.com/facebook/react/blob/769b1f270e1251d9dbdce0fcbd9e92e502d059b8/packages/shared/getComponentName.js
 *
 * @param {React.ReactType} Component
 * @returns {string | undefined}
 */
export default (Component) => {
    if (Component == null) {
        return undefined;
    }

    if (typeof Component === 'string') {
        return Component;
    }

    if (typeof Component === 'function') {
        return getFunctionComponentName(Component, 'Component');
    }

    if (typeof Component === 'object') {
        switch (Component.$$typeof) {
            case ForwardRef():
                return getWrappedName(Component, Component.render, 'ForwardRef');
            default:
                return undefined;
        }
    }

    return undefined;
};
