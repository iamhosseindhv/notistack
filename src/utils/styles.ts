import { css, CSSAttribute } from 'goober';

export function makeStyles<S extends { [key: string]: CSSAttribute }, K extends keyof S>(
    styles: S
): { [key in K]: string } {
    return Object.entries(styles).reduce(
        (acc, [key, value]) => ({
            ...acc,
            [key]: css(value),
        }),
        {} as { [key in K]: string }
    );
}

export const ComponentClasses = {
    SnackbarContainer: 'notistack-SnackbarContainer',
    Snackbar: 'notistack-Snackbar',
    CollapseWrapper: 'notistack-CollapseWrapper',
    MuiContent: 'notistack-MuiContent',
    MuiContentVariant: (variant: string) => `notistack-MuiContent-${variant}`,
};
