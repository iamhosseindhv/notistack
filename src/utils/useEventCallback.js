/**
 * @link https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/utils/useEventCallback.js
 */
import * as React from 'react';

const useEnhancedEffect = typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect;

export default function useEventCallback(fn) {
    const ref = React.useRef(fn);
    useEnhancedEffect(() => {
        ref.current = fn;
    });
    return React.useCallback((...args) => (0, ref.current)(...args), []);
}
