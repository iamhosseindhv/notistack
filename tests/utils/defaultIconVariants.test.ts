import { it, expect } from 'vitest';
import { customRenderer } from '../test-utils';

import { default as defaultIconVariants } from '../../src/utils/defaultIconVariants';

it('should render icon correctly: <CheckIcon />', () => {
  const checkIcon = defaultIconVariants.success;

  const { asFragment } = customRenderer(checkIcon);
  expect(asFragment()).toMatchSnapshot();
})

it('should render icon correctly: <WarningIcon />', () => {
  const warningIcon = defaultIconVariants.warning;
  
  const { asFragment } = customRenderer(warningIcon);
  expect(asFragment()).toMatchSnapshot();
})

it('should render icon correctly: <ErrorIcon />', () => {
  const errorIcon = defaultIconVariants.error;

  const { asFragment } = customRenderer(errorIcon);
  expect(asFragment()).toMatchSnapshot();
})

it('should render icon correctly: <InfoIcon />', () => {
  const infoIcon = defaultIconVariants.info;

  const { asFragment } = customRenderer(infoIcon);
  expect(asFragment()).toMatchSnapshot();
})
