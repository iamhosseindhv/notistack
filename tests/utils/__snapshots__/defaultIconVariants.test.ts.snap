// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render icon correctly: <CheckIcon /> 1`] = `
<DocumentFragment>
  <svg
    focusable="false"
    style="font-size: 20px; margin-inline-end: 8px; user-select: none; width: 1em; height: 1em; display: inline-block; fill: currentColor; flex-shrink: 0;"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41
        10.59L10 14.17L17.59 6.58L19 8L10 17Z"
    />
  </svg>
</DocumentFragment>
`;

exports[`should render icon correctly: <ErrorIcon /> 1`] = `
<DocumentFragment>
  <svg
    focusable="false"
    style="font-size: 20px; margin-inline-end: 8px; user-select: none; width: 1em; height: 1em; display: inline-block; fill: currentColor; flex-shrink: 0;"
    viewBox="0 0 24 24"
  >
    <path
      d="M12,2C17.53,2 22,6.47 22,12C22,17.53 17.53,22 12,22C6.47,22 2,17.53 2,12C2,
        6.47 6.47,2 12,2M15.59,7L12,10.59L8.41,7L7,8.41L10.59,12L7,15.59L8.41,17L12,
        13.41L15.59,17L17,15.59L13.41,12L17,8.41L15.59,7Z"
    />
  </svg>
</DocumentFragment>
`;

exports[`should render icon correctly: <InfoIcon /> 1`] = `
<DocumentFragment>
  <svg
    focusable="false"
    style="font-size: 20px; margin-inline-end: 8px; user-select: none; width: 1em; height: 1em; display: inline-block; fill: currentColor; flex-shrink: 0;"
    viewBox="0 0 24 24"
  >
    <path
      d="M13,9H11V7H13M13,17H11V11H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,
        0 22,12A10,10 0 0,0 12,2Z"
    />
  </svg>
</DocumentFragment>
`;

exports[`should render icon correctly: <WarningIcon /> 1`] = `
<DocumentFragment>
  <svg
    focusable="false"
    style="font-size: 20px; margin-inline-end: 8px; user-select: none; width: 1em; height: 1em; display: inline-block; fill: currentColor; flex-shrink: 0;"
    viewBox="0 0 24 24"
  >
    <path
      d="M13,14H11V10H13M13,18H11V16H13M1,21H23L12,2L1,21Z"
    />
  </svg>
</DocumentFragment>
`;
